server/state/in_memory_state.go

```diff
@@ -0,0 +1,460 @@
+package state
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/docker/docker/pkg/registrar"
+	"github.com/docker/docker/pkg/truncindex"
+	"github.com/kubernetes-incubator/cri-o/oci"
+	"github.com/kubernetes-incubator/cri-o/server/sandbox"
+)
+
+// TODO: make operations atomic to greatest extent possible
+
+// InMemoryState is an in-memory state store, suitable for use when no other
+// programs are expected to interact with the server
+type InMemoryState struct {
+	lock         sync.Mutex
+	sandboxes    map[string]*sandbox.Sandbox
+	containers   oci.ContainerStorer
+	podNameIndex *registrar.Registrar
+	podIDIndex   *truncindex.TruncIndex
+	ctrNameIndex *registrar.Registrar
+	ctrIDIndex   *truncindex.TruncIndex
+}
+
+// NewInMemoryState creates a new, empty server state
+func NewInMemoryState() RuntimeStateStorer {
+	state := new(InMemoryState)
+	state.sandboxes = make(map[string]*sandbox.Sandbox)
+	state.containers = oci.NewMemoryStore()
+	state.podNameIndex = registrar.NewRegistrar()
+	state.podIDIndex = truncindex.NewTruncIndex([]string{})
+	state.ctrNameIndex = registrar.NewRegistrar()
+	state.ctrIDIndex = truncindex.NewTruncIndex([]string{})
+
+	return state
+}
+
+// AddSandbox adds a sandbox and any containers in it to the state
+func (s *InMemoryState) AddSandbox(sandbox *sandbox.Sandbox) (err error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if sandbox == nil {
+		return fmt.Errorf("nil passed as sandbox to AddSandbox")
+	}
+
+	if _, exist := s.sandboxes[sandbox.ID()]; exist {
+		return fmt.Errorf("sandbox with ID %v already exists", sandbox.ID())
+	}
+
+	// We shouldn't share ID with any containers
+	// Our pod infra container will share our ID and we don't want it to conflict with anything
+	if ctrCheck := s.containers.Get(sandbox.ID()); ctrCheck != nil {
+		return fmt.Errorf("requested sandbox ID %v conflicts with existing container ID", sandbox.ID())
+	}
+
+	if err := s.addSandboxMappings(sandbox); err != nil {
+		return err
+	}
+	defer func() {
+		if err != nil {
+			if err2 := s.deleteSandboxMappings(sandbox); err2 != nil {
+				logrus.Errorf("Error removing mappings for incompletely-added sandbox %s: %v", sandbox.ID(), err)
+			}
+		}
+	}()
+
+	// If there are containers in the sandbox add them to the mapping
+	addedCtrs := make([]*oci.Container, 0, len(sandbox.Containers()))
+	for _, ctr := range sandbox.Containers() {
+		if err := s.addContainerMappings(ctr, true); err != nil {
+			return fmt.Errorf("error adding container %v mappings in sandbox %v", ctr.ID(), sandbox.ID())
+		}
+		addedCtrs = append(addedCtrs, ctr)
```

Comment:
This doesn't work right. You're allocating the array and filling it up here, but the defer is not called yet, so if you get an error adding one of the containers, the previously added ones will not get cleaned up.

Also, i think all this array complexity is unnecessary, just call defer once per added container.